components/ws-proxy/pkg/proxy/metrics.go

```diff
@@ -0,0 +1,186 @@
+			Name: "ws_proxy_server_websocket_connection_attempts_total",
+			Help: "Total number of incoming WebSocket connection attempts",
+		}, []string{"resource", "status"}),
+	}
+	clientMetrics = &httpMetrics{
+		requestsTotal: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "ws_proxy_client_http_requests_total",
+			Help: "Total number of outgoing HTTP requests",
+		}, []string{"method", "resource", "status"}),
+		requestsDuration: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Name:    "ws_proxy_client_http_requests_duration_seconds",
+			Help:    "Duration of outgoing HTTP requests in seconds",
+			Buckets: []float64{.005, .025, .05, .1, .5, 1, 2.5, 5, 30, 60, 120, 240, 600},
+		}, []string{"method", "resource", "status"}),
+		webSocketConnectionAttemptsTotal: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "ws_proxy_client_websocket_connection_attempts_total",
+			Help: "Total number of outgoing WebSocket connection attempts",
+		}, []string{"resource", "status"}),
+	}
+)
+
+func init() {
+	metrics.Registry.MustRegister(serverMetrics, clientMetrics)
+}
+
+func (m *httpMetrics) observe(req *http.Request, resource string, start time.Time, statusCode int) {
+	status := fmt.Sprintf("%d", statusCode)
+
+	webSocketUpgade := strings.ToLower(req.Header.Get("Upgrade")) == "websocket" &&
+		strings.ToLower(req.Header.Get("Connection")) == "upgrade"
+
+	if webSocketUpgade {
+		m.webSocketConnectionAttemptsTotal.WithLabelValues(resource, status).Inc()
+	} else {
+		m.requestsTotal.WithLabelValues(req.Method, resource, status).Inc()
+		m.requestsDuration.WithLabelValues(req.Method, resource, status).Observe(time.Since(start).Seconds())
+	}
+}
+
+type clientMetricsTransport struct {
+	transport http.RoundTripper
+}
+
+type contextKey int
+
+var (
+	resourceKey = contextKey(0)
+)
+
+func withResource(r *http.Request, resource string) *http.Request {
+	ctx := context.WithValue(r.Context(), resourceKey, resource)
+	return r.WithContext(ctx)
+}
+
+func getResource(r *http.Request) string {
+	if v := r.Context().Value(resourceKey); v != nil {
+		if resource, ok := v.(string); ok {
+			return resource
+		}
+	}
+	return ""
+}
+
+func (t *clientMetricsTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	start := time.Now()
+	resp, err := t.transport.RoundTrip(req)
+
+	statusCode := 0
+	if resp != nil {
+		statusCode = resp.StatusCode
+	}
+
+	resource := getResource(req)
+	if resource == "" {
+		resource = "unknown"
+		log.WithField("URL", req.URL).Warn("server metrics: cannot determine resource")
+	}
+	clientMetrics.observe(req, resource, start, statusCode)
+
+	return resp, err
+}
+
+func trackServerMetrics(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		start := time.Now()
+
+		// negroni ensures that the wrapper implements all necessary interfaces
+		rw := negroni.NewResponseWriter(w)
```

Comment:
It is very hard to get status using http middleware. One has to implement all interfaces to capture properly. negroni seems to be popular and fixes it. Any concerns using it?

Alternatively we can only collect outgoing metrics, but it won't give a complete picture, since ws-proxy is not exactly direct mapping between incoming and outgoing requests.